File: codigo.pas
CONST n = ...; (* numero de vertices *)
TYPE MATRIZ = ARRAY [1..n],[1..n] OF CARDINAL;

PROCEDURE GrafosIguales(k:CARDINAL);
VAR vertice:CARDINAL; (*vertice que indica la opcion en curso*) BEGIN
vertice:=0; 
REPEAT
INC(vertice);
X[k]:=vertice;
IF Valido(k) THEN
IF k<n THEN
GrafosIguales(k+1) ELSE
exito:=TRUE END
END
UNTIL (X[k]=n) OR exito; END GrafosIguales;


PROCEDURE Valido(k:CARDINAL):BOOLEAN; VAR i:CARDINAL;
BEGIN
FOR i:=1 TO k-1 DO (* no pueden repetirse elementos *) IF X[i]=X[k] THEN
RETURN FALSE END
END;
IF NumArcos(L1,k)<>NumArcos(L2,X[k]) THEN(* mismo num. arcos *) RETURN FALSE
END;
FOR i:=1 TO k-1 DO (* mismas conexiones *)
IF (L2[X[i],X[k]]<>L1[i,k]) OR (L2[X[k],X[i]]<>L1[k,i]) THEN RETURN FALSE
END END;
RETURN TRUE;
END Valido;


PROCEDURE NumArcos(VAR L:MATRIZ;k:CARDINAL):CARDINAL; VAR i,suma:CARDINAL;
BEGIN
suma:=0;
FOR i:=1 TO n DO
IF ((i<>k) AND (L[i,k]<MAX(CARDINAL))) THEN
INC(suma)
END;
IF ((i<>k) AND (L[k,i]<MAX(CARDINAL))) THEN
INC(suma) END;
END;
RETURN suma;
END NumArcos;

File: codigo.py
n = 4  # Valor adecuado de n
X = [0] * n  # Se inicializa X como una lista de ceros de longitud n
L1 = [[0] * n for _ in range(n)]  # Matriz L1
L2 = [[0] * n for _ in range(n)]  # Matriz L2
exito = False

def GrafosIguales(k, index):
    global exito
    if index == n:
        exito = True
        return
    for vertice in range(1, n + 1):
        X[k] = vertice
        if Valido(k):
            GrafosIguales(k + 1, index + 1)

def Valido(k):
    for i in range(1, k):
        if X[i] == X[k]:
            return False
    if NumArcos(L1, k) != NumArcos(L2, X[k]):
        return False
    for i in range(1, k):
        if (L2[X[i] - 1][X[k] - 1] != L1[i - 1][k - 1]) or (L2[X[k] - 1][X[i] - 1] != L1[k - 1][i - 1]):
            return False
    return True

def NumArcos(L, k):
    suma = 0
    for i in range(n):
        if (i != k) and (L[i][k - 1] < float('inf')):
            suma += 1
        if (i != k) and (L[k - 1][i] < float('inf')):
            suma += 1
    return suma

GrafosIguales(0, 0)
if exito:
    print("Los grafos son iguales")
else:
    print("Los grafos no son iguales")